torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -628,6 +631,71 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
 
         return outputs
 
+    @torch.no_grad()
+    def _write_back_current_shard(self):
+        for p in self.params:
+            if not p._is_sharded:
+                pass
+            chunks = p._full_param_padded.chunk(self.world_size)
+            if len(chunks) < (self.rank + 1):
+                # if there are not enough chunks, there's nothing to writeback here
+                pass
+        else:
+            chunk = chunks[self.rank]
+            p._local_shard.copy_(chunk)
+
+    @contextlib.contextmanager
+    def _summon_full_params(self, recurse: bool = True, volatile: bool = False) -> Generator:
+        """
+        A context manager to expose full params for the current FSDP instance.
+        Can be useful *after* forward/backward for a model to get the params for
+        additional processing or checking.
+        .. note:: This can be used on inner FSDPs.
+        .. note:: This can *not* be used within a forward or backward pass. Nor
+            can forward and backward be started from within this context.
```

Comment:
is it worth adding a test case to verify we throw an appropriate error in this case (or file a follow up issue to do so, doesn't have to be in this PR)?

Reply:
good point. the test I added is a bit wacky in how it allows `_summon_full_params` to be called from forward, but it does the job.